### PR TITLE
Add support for SELECT command

### PR DESCRIPTION
**Pull request by tevino**

### CODE REVIEW

**Comment by tremez** [comment]
Thank you !
Waiting, when it will appear in master

File: src/command.c [hunk 1]
```diff
@@ -672,6 +674,27 @@ int cmd_quit(struct command *cmd)
...
+    ASSERT_TYPE(data, REP_ARRAY);
+    ASSERT_ELEMENTS(data->elements == 2, data);
+
+    struct redis_data *db_data = &data->element[1];
+    ASSERT_TYPE(db_data, REP_STRING);
+
+    struct pos_array *db_str = &db_data->pos;
+    // TODO: Implement pos_to_long then use it here, someday.
+    if (db_str->str_len == 1 && db_str->items->str[0] == '0') {
+        conn_add_data(cmd->client, (uint8_t*)rep_ok, 5,
```
Comment (doyoubi) [inline src/command.c]: I think the overhead of `str_len` is acceptable.

File: src/command.c [hunk 2]
```diff
@@ -672,6 +674,27 @@ int cmd_quit(struct command *cmd)
...
+{
+    ASSERT_TYPE(data, REP_ARRAY);
+    ASSERT_ELEMENTS(data->elements == 2, data);
+
+    struct redis_data *db_data = &data->element[1];
+    ASSERT_TYPE(db_data, REP_STRING);
+
+    struct pos_array *db_str = &db_data->pos;
+    // TODO: Implement pos_to_long then use it here, someday.
+    if (db_str->str_len == 1 && db_str->items->str[0] == '0') {
```
Comment (maralla) [inline src/command.c]: The comparison is not very suitable here. If the command `select 000` is issued, `OK` should be returned.

Comment (tevino) [inline src/command.c]: Yes, are you suggesting `pos_to_long`?

File: src/command.c [hunk 3]
```diff
@@ -672,6 +674,25 @@ int cmd_quit(struct command *cmd)
     return CORVUS_OK;
 }
 
+int cmd_select(struct command *cmd, struct redis_data *data)
+{
+    ASSERT_TYPE(data, REP_ARRAY);
+    ASSERT_ELEMENTS(data->elements == 2, data);
+
+    struct redis_data *db_data = &data->element[1];
```
Comment (doyoubi) [inline src/command.c]: The second argument of `SELECT` is the index of database. `db_data` is not quite suitable here.

File: src/parser.c [hunk 4]
```diff
@@ -559,3 +559,23 @@ size_t pos_to_str_with_limit(struct pos_array *pos, uint8_t *str, size_t limit)
     }
     return len;
 }
+
+int pos_is_zero(struct pos_array *pos)
```
Comment (doyoubi) [inline src/parser.c]: Redis uses `strtoll` which is available in C99 for error handling. Both of us just missed it!

File: src/parser.c [hunk 5]
```diff
@@ -559,3 +559,23 @@ size_t pos_to_str_with_limit(struct pos_array *pos, uint8_t *str, size_t limit)
...
+{
+    int length = pos->str_len;
+    if (length <= 0) {
+        LOG(ERROR, "pos_is_zero: string length %d <= 0", length);
+        return CORVUS_ERR;
+    }
+
+    int len = 0;
+    for (size_t i = 0; len < length; i++) {
```
Comment (doyoubi) [inline src/parser.c]: `length` here should be `pos->pos_len`. See `cmd_get_map_key` for how to iterate `pos_array`.

File: src/parser.c (hunk 4 above)
Comment (tevino) [inline src/parser.c]: To make use of `strtoll`, a `pos_array` must be converted into a string, too expensive for an operation like this.

File: src/parser.c (hunk 5 above)
Comment (tevino) [inline src/parser.c]: But `pos_to_str` uses `pos->str_len` instead of `pos->pos_len`.

File: src/command.c (hunk 3 above)
Comment (tevino) [inline src/command.c]: Well it's the index of **db**, I think `db` tells more than `index` here.

File: src/parser.c (hunk 5 above)
Comment (doyoubi) [inline src/parser.c]: `pos_array` is used to represent only one string, consisting of split sub string stored in `items`
``` C
struct pos_array {
    struct pos *items;
    int str_len;                    // str_len == sum(items.len)
    int pos_len;                  // pos_len == len(items)
    int max_pos_size;       // max_pos_size is the actual memory size allocated
};
```

File: src/parser.c (hunk 4 above)
Comment (maralla) [inline src/parser.c]: If we try our best to conform to redis, we should consider this:

```
> select hello
(error) ERR invalid DB index
```

Comment (tevino) [inline src/parser.c]: Agree, I've thought about that.

The error messages are completely different in some cases:
```
Redis:
> select
(error) ERR wrong number of arguments for 'select' command

Corvus:
> select
(error) ERR Proxy fail to forward command
```

I'm afraid it's too much work to get them exactly the same in this PR.

And if nothing goes wrong, do we have to do this?

File: src/parser.c (hunk 5 above)
Comment (tevino) [inline src/parser.c]: Thanks for the hint, what about now?

File: src/parser.c (hunk 4 above)
Comment (doyoubi) [inline src/parser.c]: I think supporting ```"SELECT is not allowed in cluster mode"``` is enough.

Comment (maralla) [inline src/parser.c]: Yeah it's enough.

**Pull request merged**